Sources/OuterCore/IRVerifier.swift

```diff
@@ -0,0 +1,165 @@
+    let qualifier = expected > got ? "few" : "many"
+    let kind = function is Continuation ? "function" : "value"
+    return .init(.error,
+                 """
+                 Graph IR: too \(qualifier) arguments to Graph IR \(kind) \
+                 '\(function.name)'; expected \(expected), got \(got)
+                 """)
+  }
+  static func typeMismatch(expected: Type, got: Type) -> Diagnostic.Message {
+    return .init(.error,
+                 """
+                 Graph IR: type mismatch (expected '\(name(for: expected))', \
+                 got '\(name(for: got))')
+                 """)
+  }
+
+  static func unsubstitutedType(
+    _ type: ParameterizedType) -> Diagnostic.Message {
+    return .init(.error,
+                 """
+                 Graph IR: type \(name(for: type)) must have all parameters \
+                 substituted
+                 """)
+  }
+}
+
+public final class IRVerifier {
+  let module: Module
+  let engine: DiagnosticEngine
+
+  var currentScopedValues = Set<Value>()
+
+  public init(module: Module, engine: DiagnosticEngine) {
+    self.module = module
+    self.engine = engine
+  }
+
+  func withScope<T>(_ actions: () -> T) -> T {
+    let oldScope = currentScopedValues
+    defer {
+      currentScopedValues = oldScope
+    }
+    return actions()
+  }
+
+  func valueIsKnown(_ value: Value) -> Bool {
+    switch value {
+    case let type as RecordType:
+      return module.knownRecordTypes.contains(type)
+    case let type as FunctionType:
+      return module.knownFunctionTypes.contains(type)
+    case let type as DataType:
+      return module.knownDataTypes.contains(type)
+    case let type as BottomType:
+      return module.bottomType === type
+    case let type as TypeMetadataType:
+      return module.metadataType === type
+    case let type as TypeType:
+      return module.typeType === type
+    case let type as ArchetypeType:
+      return valueIsKnown(type.parent)
+    case let type as SubstitutedType:
+      guard valueIsKnown(type.substitutee) else { return false }
+      for (arch, subst) in type.substitutions {
+        guard valueIsKnown(arch) else { return false }
+        guard valueIsKnown(subst) else { return false }
+      }
+      return true
+    default:
+      return currentScopedValues.contains(value)
+    }
+  }
+
+  func verifyType(_ type: Value) -> Bool {
+    guard valueIsKnown(type) else {
+      engine.diagnose(.unknownType(type))
+      return false
+    }
+    switch type {
+    case let type as ParameterizedType:
+      engine.diagnose(.unsubstitutedType(type))
+      return false
+    default:
+      break
+    }
+    return true
+  }
+
+  @discardableResult
+  public func verify() -> Bool {
+    var allValid = true
+    for continuation in module.continuations {
+      allValid &= verify(continuation)
+    }
+    return allValid
+  }
+
+  func verify(_ continuation: Continuation) -> Bool {
+    return withScope {
```

Comment:
We had a discussion offline about reworking this to just crash.  Diagnosing in the verifier is Nice ™, but we're already implementing 2-and-a-half languages 😆

Reply:
Yeah, and actually we'll have more context if we crash, because of [PrettyStackTrace](https://github.com/llvm-swift/PrettyStackTrace)